### PR TITLE
add assembly compilation support

### PR DESCRIPTION
**Pull request by weissi**

### CODE REVIEW

File: Sources/PackageModel/Manifest.swift [hunk 1]
```diff
@@ -15,12 +15,14 @@ import Utility
 public enum ManifestVersion: String, Codable {
     case v4 = "4"
     case v4_2 = "4_2"
+    case v5 = "5"
```
Comment (aciidb0mb3r) [inline Sources/PackageModel/Manifest.swift]: I am actually currently working on adding support for manifest v5. I should be done soon, let's wait until I merge my changes as there are some subtle things to take care of when adding a new manifest version.

File: Sources/PackageModel/Sources.swift [hunk 2]
```diff
@@ -51,6 +51,9 @@ public enum SupportedLanguageExtension: String {
     case cc
     case cpp
     case cxx
+    /// Assembly
+    case s
+    case S
```
Comment (neonichu) [inline Sources/PackageModel/Sources.swift]: We should also add .asm here, it is a supported extension for assembly in Xcode.

File: Sources/PackageModel/Manifest.swift (hunk 1 above)
Comment (neonichu) [inline Sources/PackageModel/Manifest.swift]: I think @aciidb0mb3r will make a change soon to support v5, this PR should be based on that.

File: Sources/PackageModel/Sources.swift (hunk 2 above)
Comment (ddunbar) [reply Sources/PackageModel/Sources.swift]: I don’t think we have to follow precedent with Xcode or Clang. We already break with Clang and I think that is good for consistency. I haven’t found “asm” to be widespread, but also haven’t don’t an audit. I’d prefer we start conservative, or do an audit to decide.

File: Sources/PackageModel/Manifest.swift (hunk 1 above)
Comment (weissi) [reply Sources/PackageModel/Manifest.swift]: @aciidb0mb3r thanks, now that your change has landed, I removed this

File: Sources/PackageModel/Sources.swift (hunk 2 above)
Comment (weissi) [reply Sources/PackageModel/Sources.swift]: @neonichu / @ddunbar / @jakepetroules what's the verdict here? I'd be in favour if just leaving .s and .S (and no .asm) but I really don't mind

File: Sources/PackageModel/Manifest.swift (hunk 1 above)
Comment (weissi) [reply Sources/PackageModel/Manifest.swift]: indeed, thanks, done

File: Sources/PackageModel/Sources.swift (hunk 2 above)
Comment (aciidb0mb3r) [reply Sources/PackageModel/Sources.swift]: We can add .asm later if needed. A restrictive start is better than an open-ended one anyway.

File: Sources/PackageModel/Sources.swift [hunk 3]
```diff
@@ -67,15 +70,25 @@ public enum SupportedLanguageExtension: String {
         SupportedLanguageExtension.stringSet(mm, cc, cpp, cxx)
     }()
 
-    /// Returns a set of valid c family extensions.
-    public static var cFamilyExtensions: Set<String> = {
-        cExtensions.union(cppExtensions)
+    public static var assemblyExtensions: Set<String> = {
```
Comment (aciidb0mb3r) [inline Sources/PackageModel/Sources.swift]: nit: doc comment plz

File: Tests/BuildTests/BuildPlanTests.swift [hunk 4]
```diff
@@ -440,6 +441,39 @@ final class BuildPlanTests: XCTestCase {
...
+                                        Manifest.createV4Manifest(
+                                            name: "Pkg",
+                                            path: "/Pkg",
+                                            url: "/Pkg",
+                                            manifestVersion: .v5,
+                                            targets: [
+                                                TargetDescription(name: "exe", dependencies: ["lib"]),
+                                                TargetDescription(name: "lib", dependencies: []),
+                                                ]),
+                                        ]
```
Comment (aciidb0mb3r) [inline Tests/BuildTests/BuildPlanTests.swift]: Can you indent this like rest of the file?

File: Tests/BuildTests/BuildPlanTests.swift [hunk 5]
```diff
@@ -440,6 +441,39 @@ final class BuildPlanTests: XCTestCase {
...
+            "/Pkg/Sources/exe/main.swift",
+            "/Pkg/Sources/lib/lib.c",
+            "/Pkg/Sources/lib/lib.S",
+            "/Pkg/Sources/lib/include/lib.h"
+        )
+
+        let diagnostics = DiagnosticsEngine()
+        let graph = loadPackageGraph(root: "/Pkg", fs: fs, diagnostics: diagnostics,
+                                     manifests: [
+                                        Manifest.createV4Manifest(
```
Comment (aciidb0mb3r) [inline Tests/BuildTests/BuildPlanTests.swift]: I added a `Manifest.createManifest` method for generic manifests

File: Sources/PackageModel/Sources.swift (hunk 3 above)
Comment (weissi) [reply Sources/PackageModel/Sources.swift]: done

File: Tests/BuildTests/BuildPlanTests.swift (hunk 4 above)
Comment (weissi) [reply Tests/BuildTests/BuildPlanTests.swift]: done

File: Tests/BuildTests/BuildPlanTests.swift (hunk 5 above)
Comment (weissi) [reply Tests/BuildTests/BuildPlanTests.swift]: awesome, thanks, changed

**Comment by weissi** [comment]
@swift-ci please test

File: Tests/BuildTests/BuildPlanTests.swift [hunk 6]
```diff
@@ -627,6 +661,76 @@ final class BuildPlanTests: XCTestCase {
       #endif
     }
 
+    func testAsmOnlyModuleV4_2Manifest() throws {
```
[the remark this answers is not in the document]
Comment (weissi) [reply Tests/BuildTests/BuildPlanTests.swift]: @aciidb0mb3r they currently use `BuildPlan*` which is a `private struct` in `PackageBuilderTests.swift`

[the remark this answers is not in the document]
Comment (aciidb0mb3r) [reply Tests/BuildTests/BuildPlanTests.swift]: oh, I didn't mean move as-is. Basically recreate such package structure in those tests and then load  the package and check if the assembly files are present in package's sources or not.

File: Sources/PackageModel/Sources.swift (hunk 2 above)
Comment (ddunbar) [reply Sources/PackageModel/Sources.swift]: I prefer .s and .S only

File: Tests/BuildTests/BuildPlanTests.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (weissi) [reply Tests/BuildTests/BuildPlanTests.swift]: done

**Comment by weissi** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

**Pull request merged**